### PR TITLE
Add Span<T>

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

File: mcs/class/Makefile [hunk 1]
```diff
@@ -264,7 +264,10 @@ net_4_x_dirs := \
 	System.Web.Http \
 	System.Web.Http.SelfHost \
 	System.Web.Http.WebHost \
-	System.Runtime.InteropServices.RuntimeInformation
+	System.Runtime.InteropServices.RuntimeInformation	\
+	System.Runtime.CompilerServices.Unsafe	\
+	System.Memory	\
```
Comment (alexrp) [inline mcs/class/Makefile]: That trailing `\` should probably not be there.

**Comment by kumpera** [comment]
@mareks That's not correct. Both assemblies must be provided by the runtime as they depend on runtime internals.

CoreCLR have all those types on corlib with a different implementation than CoreFX.

I was trying to get a baseline to build against, but given using xunit turned out to be a dumper fire, I agree that there's little value into using CoreFX at all.

**Comment by marek-safar** [comment]
We need to fix this in CoreFX otherwise we end up again in that denied loading mess.

**Comment by marek-safar** [comment]
XUnit stuff is being worked on but not yet part of mono build

**Comment by kumpera** [comment]
If CoreCLR gets to fix this in their mscorlib so must we be able to do the same.

**Comment by marek-safar** [comment]
I'll check how what's the runtime api surface but CoreCLR mscorlib is a private assembly so it should not be hard to inject something between.

**Comment by marek-safar** [comment]
I looked into it more deeply and here is what I suggest us to do
1. Close this PR and submit the PR for ilasm changes which is the valuable bit
2. Create a new assembly called System.Corlib.Private
3. Link CoreRT/CoreFX Span<T> implementation and implement mono version of SpanHelper.

That way we won't interfere with any nugets and keep our maintenance work to minimum.

I can set this up for you if you want so you can "just" implement the runtime logic

**Comment by jkotas** [comment]
> That way we won't interfere with any nugets 

Is it still the plan to not create Mono TFM? If it is still the plan, I except that you will collide with System.Memory for full .NET Framework and will need to add System.Memory to the Mono nuget blacklist.

**Comment by marek-safar** [comment]
@jkotas Mono TFM was rejected. I don't know if System.Memory will appear as part of .NET Framework or not but I'd prefer us not to build and deploy System.Memory if we don't have to.

**Comment by kumpera** [comment]
Hi Marek,

I agree with most of your plan. But as it turns out, the version on corefx is not the efficient 2 fields one. That version is on coreclr. The end goal is to get the coreclr one to work under mono.

I'll throw the ilasm bits in a separate PR and I'd love if you scaffold System.Corlib.Private. I'll port as many tests as reasonable from xunit.

**Comment by marek-safar** [comment]
No, reason to port xunit tests. I have PR to run any xunit corefx tests on mono (blocked by RE for now :-(

**Comment by jkotas** [comment]
> the version on corefx is not the efficient 2 fields one

Correct. Nuget will give you this version for .NET Framework. If you want a different one for Mono, you will need to override it using the blacklist.

**Pull request closed without merging**